File: python/app/schemas/marketplace.py
from pydantic import BaseModel, Field
from typing import Literal

from app.models.marketplace import ContactVisibility, CompensationType, OpportunityStatus


class SkillCreate(BaseModel):
    name: str = Field(min_length = 1, max_length = 120)
    category: str | None = Field(default = None, max_length = 120)


class SkillOut(BaseModel):
    id: int
    name: str
    category: str | None = None

    class Config:
        from_attributes = True


class WorkerProfileUpdate(BaseModel):
    bio: str | None = None
    experience_summary: str | None = None
    wants_paid: bool | None = None
    wants_volunteer: bool | None = None
    exact_latitude: float | None = None
    exact_longitude: float | None = None
    public_latitude: float | None = None
    public_longitude: float | None = None
    city: str | None = None
    region: str | None = None
    contact_visibility: ContactVisibility | None = None
    is_available: bool | None = None


class WorkerProfileOut(BaseModel):
    user_id: int
    username: str
    bio: str | None = None
    experience_summary: str | None = None
    wants_paid: bool
    wants_volunteer: bool
    city: str | None = None
    region: str | None = None
    public_latitude: float | None = None
    public_longitude: float | None = None
    exact_latitude: float | None = None
    exact_longitude: float | None = None
    contact_visibility: ContactVisibility
    is_available: bool
    skills: list[SkillOut] = []


class WorkerSkillsUpdate(BaseModel):
    skill_ids: list[int]


class ContactChannelCreate(BaseModel):
    channel_type: str = Field(min_length = 1, max_length = 50)
    channel_value: str = Field(min_length = 1, max_length = 255)
    visibility: ContactVisibility = ContactVisibility.gov_only
    is_primary: bool = False


class ContactChannelOut(BaseModel):
    id: int
    channel_type: str
    channel_value: str
    visibility: ContactVisibility
    is_primary: bool

    class Config:
        from_attributes = True


class EmployerProfileUpdate(BaseModel):
    organization_name: str | None = Field(default = None, max_length = 160)
    organization_description: str | None = None


class EmployerProfileOut(BaseModel):
    user_id: int
    username: str
    organization_name: str
    organization_description: str | None = None
    is_government_service: bool
    is_verified: bool


class OpportunityCreate(BaseModel):
    title: str = Field(min_length = 1, max_length = 200)
    description: str = Field(min_length = 1)
    compensation_type: CompensationType
    budget_note: str | None = Field(default = None, max_length = 120)
    latitude: float
    longitude: float
    city: str | None = Field(default = None, max_length = 120)
    region: str | None = Field(default = None, max_length = 120)
    skill_ids: list[int] = []


class OpportunityOut(BaseModel):
    id: int
    employer_profile_id: int
    employer_id: int
    title: str
    description: str
    compensation_type: CompensationType
    budget_note: str | None = None
    latitude: float
    longitude: float
    city: str | None = None
    region: str | None = None
    status: OpportunityStatus
    skills: list[SkillOut] = []


class ApplicationCreate(BaseModel):
    message: str | None = None


class ApplicationOut(BaseModel):
    id: int
    opportunity_id: int
    worker_profile_id: int
    worker_id: int
    status: str
    message: str | None = None

    class Config:
        from_attributes = True

File: frontend/src/services/marketplace.ts
import apiClient, { type RequestConfig } from './apiClient';
import type { AccessToken } from '@/types/ids';
import { AppApiPaths } from '@/types/types';

export interface WorkerProfile {
	user_id: number;
	username: string;
	bio?: string;
	experience_summary?: string;
	wants_paid: boolean;
	wants_volunteer: boolean;
	city?: string;
	region?: string;
	public_latitude?: number;
	public_longitude?: number;
	exact_latitude?: number;
	exact_longitude?: number;
	contact_visibility?: string;
	is_available?: boolean;
	skills: Array<{ id: number; name: string }>;
}

export interface SearchWorkersParams {
	skill_ids?: number[];
	wants_paid?: boolean;
	wants_volunteer?: boolean;
	city?: string;
	region?: string;
	min_lat?: number;
	max_lat?: number;
	min_lng?: number;
	max_lng?: number;
}

export async function searchWorkers(
	params: SearchWorkersParams,
	token?: AccessToken
): Promise<WorkerProfile[]> {
	const queryParams = new URLSearchParams();

	if (params.skill_ids?.length) {
		params.skill_ids.forEach(id => queryParams.append('skill_ids', id.toString()));
	}
	if (params.wants_paid !== undefined) {
		queryParams.append('wants_paid', params.wants_paid.toString());
	}
	if (params.wants_volunteer !== undefined) {
		queryParams.append('wants_volunteer', params.wants_volunteer.toString());
	}
	if (params.city) {
		queryParams.append('city', params.city);
	}
	if (params.region) {
		queryParams.append('region', params.region);
	}
	if (params.min_lat !== undefined) {
		queryParams.append('min_lat', params.min_lat.toString());
	}
	if (params.max_lat !== undefined) {
		queryParams.append('max_lat', params.max_lat.toString());
	}
	if (params.min_lng !== undefined) {
		queryParams.append('min_lng', params.min_lng.toString());
	}
	if (params.max_lng !== undefined) {
		queryParams.append('max_lng', params.max_lng.toString());
	}

	const url = `marketplace/workers/search?${queryParams.toString()}`;
	const config: RequestConfig = {};

	if (token) {
		config.token = token;
	}

	return (await apiClient.get(url, config)) as WorkerProfile[];
}

export async function searchVolunteers(
	token?: AccessToken
): Promise<WorkerProfile[]> {
	return searchWorkers({ wants_volunteer: true }, token);
}
